File: Problem2.js
//Time: O(n*n), Space: O(1)
//Runs on leetcode solution accepted
var minFallingPathSum = function(matrix) {
    if(matrix.length == 0 ) return 0;
    let n = matrix.length;
    for(let i=n-2;i>=0;i--){
        for(let j=0;j<n;j++){
            if(j==0){
                matrix[i][j] = matrix[i][j] + Math.min(matrix[i+1][j],matrix[i+1][j+1] )
            }else if(j == n-1){
                matrix[i][j] = matrix[i][j] + Math.min(matrix[i+1][j],matrix[i+1][j-1] )
            }else{
                matrix[i][j] = matrix[i][j] + Math.min(matrix[i+1][j-1],Math.min(matrix[i+1][j],matrix[i+1][j+1]) )
            }
        }
    }
    let min = matrix[0][0]
    
    for(let i=1;i<n;i++){
        min = Math.min(min,matrix[0][i])
    }
    return min
};

File: problem2.cpp

// Time Complexity : O(m*n)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : yes

// we add matrix value to the minimum that we get from other row

class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        for(int i=1;i<matrix.size();i++)
        {
            for(int j=0; j<matrix[0].size();j++)
            {
                if(j==0)
                {
                    matrix[i][j] = min(matrix[i-1][j],matrix[i-1][j+1]) + matrix[i][j];
                }
                else if(j== matrix[0].size()-1)
                {
                    matrix[i][j] = min(matrix[i-1][j],matrix[i-1][j-1]) + matrix[i][j];
                }
                else
                {
                    matrix[i][j] = min(matrix[i-1][j-1],min(matrix[i-1][j],matrix[i-1][j+1])) + matrix[i][j];
                }
            }
        }
        int ans = INT_MAX;
        for(int i=0;i<matrix[0].size();i++)
        {
            ans = min(ans,matrix[matrix.size()-1][i]);
        }
        return ans;
    }
};